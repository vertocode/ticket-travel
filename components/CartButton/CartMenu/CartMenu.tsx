'use client'
import {ReactElement} from 'react'

// styles
import '@/styles/components/CartButton/CartMenu/CartMenu.scss'

// types
import type {Ticket} from "@/types/Ticket"

// components
import Divider from '@mui/material/Divider';
import CartTicketDetails from './CartTicketDetails/CartTicketDetails'
import CartTicketSubtotal from './CartTicketSubtotal/CartTicketSubtotal'
import CartTicketTotal from './CartTicketTotal/CartTicketTotal'

interface Props {
    tickets: Ticket[]
}

const CartMenu = ({ tickets }: Props): ReactElement => {
    return (
        <div className="cart-menu">
            <h3 className="title">Ingressos</h3>
            {tickets.length ? (
                <>
                    {tickets.map(ticket => (
                        <CartTicketDetails
                            ticket={ticket}
                            key={ `ticket-details-${ticket.id}` }
                        />
                    )) }
                    <Divider />
                    <CartTicketSubtotal tickets={ tickets }/>
                    <Divider />
                    <CartTicketTotal tickets={ tickets }/>
                    <div className="check-out-button">Ir para o check out</div>
                </>
            ) : (
                <div className="empty-cart-message">Não há ingressos no carrinho.</div>
            )}
        </div>
    )
}

export default CartMenu