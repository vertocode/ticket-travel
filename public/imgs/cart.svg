<svg xmlns="http://www.w3.org/2000/svg" width="24" height="23" viewBox="0 0 24 23" fill="none">
    <path d="M8.86755 21.5625C9.41492 21.5625 9.85864 21.1188 9.85864 20.5714C9.85864 20.024 9.41492 19.5803 8.86755 19.5803C8.32019 19.5803 7.87646 20.024 7.87646 20.5714C7.87646 21.1188 8.32019 21.5625 8.86755 21.5625Z" fill="white" stroke="white" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
    <path d="M19.767 21.5625C20.3143 21.5625 20.7581 21.1188 20.7581 20.5714C20.7581 20.024 20.3143 19.5803 19.767 19.5803C19.2196 19.5803 18.7759 20.024 18.7759 20.5714C18.7759 21.1188 19.2196 21.5625 19.767 21.5625Z" fill="white" stroke="white" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M0.9375 0C0.523286 0 0.1875 0.335786 0.1875 0.75C0.1875 1.16421 0.523286 1.5 0.9375 1.5H4.2871L5.15133 5.81793C5.15493 5.84188 5.15966 5.86545 5.16547 5.8886L6.82235 14.1668L6.82242 14.1672C6.94741 14.796 7.28953 15.3608 7.78892 15.7628C8.28641 16.1633 8.90839 16.3768 9.54679 16.3663H19.1669C19.8053 16.3768 20.4273 16.1633 20.9248 15.7628C21.4243 15.3607 21.7664 14.7957 21.8914 14.1668L21.8914 14.1668L21.8925 14.1612L23.4782 5.84594C23.5201 5.62638 23.4618 5.39958 23.3193 5.22736C23.1769 5.05514 22.965 4.95545 22.7415 4.95545H6.50845L5.63727 0.602807C5.56712 0.252289 5.25933 0 4.90186 0H0.9375Z" fill="white"/>
</svg>