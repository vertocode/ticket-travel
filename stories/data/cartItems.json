[
  {
    "id": "1",
    "name": "Museu de Arte Moderna",
    "location": "Rio de Janeiro, RJ",
    "image": "https://i.postimg.cc/zD6RYYc7/museu-de-arte.jpg",
    "description": "O Museu de Arte Moderna é um espaço cultural dedicado à exibição de obras de arte moderna e contemporânea, destacando-se pela diversidade de artistas e estilos.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "2",
    "name": "Torre Eiffel",
    "location": "Paris, França",
    "image": "https://i.postimg.cc/JhFkqsSN/torre-eiffel.jpg",
    "description": "A Torre Eiffel é um ícone da arquitetura mundial, situada em Paris, oferecendo uma vista deslumbrante da cidade luz a partir de seus diferentes níveis.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "3",
    "name": "Praia de Copacabana",
    "location": "Rio de Janeiro, RJ",
    "image": "https://i.postimg.cc/pXB5SGgX/praia-de-copacabana.jpg",
    "description": "A Praia de Copacabana é famosa por sua extensa faixa de areia, águas tropicais e o calçadão movimentado, sendo um dos destinos turísticos mais populares do Brasil.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "4",
    "name": "Coliseu de Roma",
    "location": "Roma, Itália",
    "image": "https://i.postimg.cc/Dz4Z7BYG/coliseu-de-roma.jpg",
    "description": "O Coliseu de Roma é um anfiteatro histórico conhecido por sua arquitetura grandiosa, palco de espetáculos e batalhas na Roma Antiga.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "5",
    "name": "Muralha da China",
    "location": "Pequim, China",
    "image": "https://i.postimg.cc/nr2V35My/muralha-da-china.jpg",
    "description": "A Grande Muralha da China é uma maravilha arquitetônica e uma das estruturas mais famosas do mundo, construída para proteger contra invasões ao longo dos séculos.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "6",
    "name": "Museu do Louvre",
    "location": "Paris, França",
    "image": "https://i.postimg.cc/PxktThDG/museu-de-louvre.jpg",
    "description": "O Museu do Louvre é o maior museu de arte do mundo, abrigando uma vasta coleção que vai desde antiguidades até obras-primas renascentistas e arte contemporânea.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "7",
    "name": "Grand Canyon",
    "location": "Arizona, EUA",
    "image": "https://i.postimg.cc/W1kL3CTs/grand-canyon.jpg",
    "description": "O Grand Canyon é uma formação geológica espetacular, conhecida por suas imponentes rochas avermelhadas esculpidas pelo rio Colorado ao longo de milhões de anos.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "8",
    "name": "Ópera de Sydney",
    "location": "Sydney, Austrália",
    "image": "https://i.postimg.cc/PJx0kFSs/opera-de-sydney.jpg",
    "description": "A Ópera de Sydney é um ícone arquitetônico mundial, conhecida por sua estrutura distintiva em forma de conchas e por sediar performances artísticas de classe mundial.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "9",
    "name": "Machu Picchu",
    "location": "Cusco, Peru",
    "image": "https://i.postimg.cc/fbqrfvw8/machu-picchu.jpg",
    "description": "Machu Picchu é uma antiga cidade inca situada nas montanhas dos Andes, reconhecida como Patrimônio Mundial da UNESCO e uma das maravilhas do mundo.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "10",
    "name": "Templo de Angkor Wat",
    "location": "Siem Reap, Camboja",
    "image": "https://i.postimg.cc/gJ0gwL83/templo-de-angkor-wat.jpg",
    "description": "O Templo de Angkor Wat é o maior templo religioso do mundo, com uma arquitetura deslumbrante e significado histórico na civilização khmer.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "11",
    "name": "Parque Nacional Yellowstone",
    "location": "Wyoming, EUA",
    "image": "https://i.postimg.cc/7ZxtvRhB/parque-naciona-yellowstone.jpg",
    "description": "O Parque Nacional de Yellowstone é famoso por suas fontes termais, gêiseres e vida selvagem diversificada, sendo o primeiro parque nacional do mundo.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "12",
    "name": "Catedral de São Basílio",
    "location": "Moscou, Rússia",
    "image": "https://i.postimg.cc/V6MW095r/catedral-de-sao-basilio.jpg",
    "description": "A Catedral de São Basílio é um ícone da arquitetura russa, conhecida por suas cúpulas coloridas e localizada na Praça Vermelha, em Moscou.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "13",
    "name": "Monte Everest",
    "location": "Himalaias, Nepal",
    "image": "https://i.postimg.cc/3NFjnKX3/monte-everest.jpg",
    "description": "O Monte Everest é a montanha mais alta do mundo, atraindo alpinistas e aventureiros de todo o mundo para desafiar seus picos cobertos de neve.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "14",
    "name": "Ilha de Páscoa",
    "location": "Rapa Nui, Chile",
    "image": "https://i.postimg.cc/kMst1hrM/ilha-de-pascoa.jpg",
    "description": "A Ilha de Páscoa é conhecida por suas estátuas de pedra gigantes, chamadas moais, que representam uma expressão única da cultura polinésia.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "15",
    "name": "Cataratas do Iguaçu",
    "location": "Foz do Iguaçu, Brasil",
    "image": "https://i.postimg.cc/bJTd5dBV/cataratas-do-iguacu.jpg",
    "description": "As Cataratas do Iguaçu são uma maravilha natural com uma série impressionante de quedas d'água na fronteira entre Brasil e Argentina.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "16",
    "name": "Palácio de Versalhes",
    "location": "Versalhes, França",
    "image": "https://i.postimg.cc/3RKx2YxX/palacio-de-versalhes.jpg",
    "description": "O Palácio de Versalhes é um magnífico palácio real conhecido por seus jardins deslumbrantes e arquitetura opulenta, localizado nos arredores de Paris.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "17",
    "name": "Parque Nacional Serengeti",
    "location": "Norte da Tanzânia",
    "image": "https://i.postimg.cc/ydQs9R5K/parque-nacional-serengeti.jpg",
    "description": "O Parque Nacional Serengeti é famoso por sua vida selvagem espetacular, incluindo a Grande Migração de gnus e zebras, criando um espetáculo impressionante.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "18",
    "name": "Monte Fuji",
    "location": "Honshu, Japão",
    "image": "https://i.postimg.cc/kMH7gxKF/monte-fuji.jpg",
    "description": "O Monte Fuji é um vulcão icônico e a montanha mais alta do Japão, oferecendo vistas deslumbrantes e sendo reverenciado na cultura japonesa.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "19",
    "name": "Praia de Anse Lazio",
    "location": "Praslin, Seychelles",
    "image": "https://i.postimg.cc/766yx6LW/praia-anse-lazio.jpg",
    "description": "Anse Lazio é uma praia paradisíaca com areias brancas e águas cristalinas, situada em uma ilha tropical no Oceano Índico.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "20",
    "name": "Monte Kilimanjaro",
    "location": "Tanzânia",
    "image": "https://i.postimg.cc/WpxtHRWv/monte-kilimanjaro.jpg",
    "description": "O Monte Kilimanjaro é a montanha mais alta da África, destacando-se por suas geleiras no topo e oferecendo uma aventura desafiadora para os alpinistas.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "21",
    "name": "Palácio de Potala",
    "location": "Lassa, Tibete",
    "image": "https://i.postimg.cc/dVY3J2cN/palacio-de-potala.jpg",
    "description": "O Palácio de Potala é uma impressionante fortaleza e palácio no Tibete, conhecido por sua arquitetura única e por ser um Patrimônio Mundial da UNESCO.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "22",
    "name": "Parque Nacional de Banff",
    "location": "Alberta, Canadá",
    "image": "https://i.postimg.cc/sxWfrX6R/parque-nacional-banff.jpg",
    "description": "O Parque Nacional de Banff é um paraíso natural com montanhas majestosas, lagos cristalinos e uma variedade abundante de vida selvagem nas Montanhas Rochosas.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  },
  {
    "id": "23",
    "name": "Teatro Colón",
    "location": "Buenos Aires, Argentina",
    "image": "https://i.postimg.cc/sDpfHMGY/teatro-colon.jpg",
    "description": "O Teatro Colón é uma joia arquitetônica e uma das principais casas de ópera do mundo, reconhecido por sua acústica impecável e performances de classe mundial.",
    "price": {
      "full": 2351.28,
      "discount": 1391.28
    },
    "rating": {
      "reviewsCount": 345,
      "value": 4.3
    },
    "createdAt": "2022-08-31T12:32:09.794Z",
    "updatedAt": "2022-08-31T12:32:09.794Z"
  }
]