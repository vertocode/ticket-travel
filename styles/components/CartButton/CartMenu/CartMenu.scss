.cart-menu {
  min-width: 300px;
  width: 440px;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  padding: 24px 24px 28px 24px;
  background-color: white;
  color: #0A1E31;
  z-index: 10;
  position: absolute;
  top: 160%;
  right: 0;

  .title {
    font-size: 1.1em;
    text-align: left;
  }

  .check-out-button {
    cursor: pointer;
    display: flex;
    justify-content: center;
    align-items: center;
    font-family: 'Inter', sans-serif;
    font-weight: 400;
    border-radius: 6px;
    padding: 16px 0;
    color: #fff;
    background-color: #0045F3;
    margin-top: 20px;
  }

  .empty-cart-message {
    font-size: 1.2em;
    color: #888;
    margin-top: 10px;
  }
}

.storybook-mode {
  position: static;
}