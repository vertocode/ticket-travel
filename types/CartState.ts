import {Ticket} from "@/types/Ticket";

export interface CartState {
    tickets: Ticket[]
}