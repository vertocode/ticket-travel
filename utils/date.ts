/**
 * Format date in brazilian format.
 *
 * @param {string} date - Date to be formatted in this format: "2022-08-31T12:32:09.794Z".
 * This is the same that API is returning.
 */
export const formatDate = (date: string) => {
	const data = new Date(date)

	return new Intl.DateTimeFormat('pt-BR', {
		day: '2-digit',
		month: '2-digit',
		year: 'numeric'
	}).format(data)
}