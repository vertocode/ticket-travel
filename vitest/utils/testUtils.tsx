import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'

import type { AppStore, RootState } from '@/lib/store'

import cartReducer from '@/lib/features/cart/cartSlice'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>
    store?: AppStore
}

export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {},
		// Automatically create a store instance if no store was passed in
		store = configureStore({ reducer: combineReducers({ cartReducer }), preloadedState }),
		...renderOptions
	}: ExtendedRenderOptions = {}
) {
	// there is a recommendation by testing-library docs to use userEvent.setup() before anything in the tests.
	// reference: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
	const user = userEvent.setup()

	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return <Provider store={store}>{children}</Provider>
	}

	// Return an object with the store and all of RTL's query functions
	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }), user }
}