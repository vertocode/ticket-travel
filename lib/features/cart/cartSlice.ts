import { createSlice } from '@reduxjs/toolkit'
import { CartState } from "@/types/CartState";

const initialState = {
    tickets: []
}

const cartSlice = createSlice({
    name: 'cart',
    initialState: initialState as CartState,
    reducers: {
        addTicket(state, action) {
            const isAlreadyAdded = state.tickets.some(ticket => ticket.id === action.payload.id)
            if (isAlreadyAdded) {
                state.tickets = state.tickets.map(ticket => {
                    if (ticket.id === action.payload.id) {
                        return {
                            ...ticket,
                            quantity: ticket.quantity + 1
                        }
                    }
                    return ticket
                })
            } else {
                state.tickets.push({ ...action.payload, quantity: 1 })
            }
        },
        deleteTicket(state, action) {
            state.tickets = state.tickets.filter(ticket => ticket.id !== action.payload)
        }
    }
})

export const { deleteTicket, addTicket } = cartSlice.actions

export default cartSlice.reducer