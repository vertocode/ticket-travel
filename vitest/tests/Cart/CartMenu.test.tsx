import { expect, test, vi } from 'vitest'
import { TicketCart } from '@/types/CartState'
import cartItems from '@/vitest/data/cartItems.json'
import CartMenu from '@/components/CartButton/CartMenu/CartMenu'
import { renderWithProviders } from '@/vitest/utils/testUtils'
import { formatDate } from '@/utils/date'
import { deleteTicket } from '@/lib/features/cart/cartSlice'

const tickets: TicketCart[] = cartItems.map(item => ({ ...item, quantity: 1 }))

const {
	container,
	getByTestId,
	getAllByTestId,
	rerender,
	getByText,
	user,
	store
} = renderWithProviders(<CartMenu tickets={[]} />)

test('should match snapshot when there are no tickets', () => {
	expect(container).toMatchSnapshot()
})

test('cart menu should be rendered with a message when there are no tickets', () => {
	expect(getByTestId('empty-cart-message').textContent).eq('Não há ingressos no carrinho.')
})

test('should match snapshot when there are tickets', () => {
	const slicedTickets = tickets.slice(0, 2)
	rerender(<CartMenu tickets={slicedTickets} />)
	expect(container).toMatchSnapshot()
})

test('cart menu should be rendered with tickets', () => {
	const slicedTickets = tickets.slice(0, 2)
	rerender(<CartMenu tickets={slicedTickets} />)
	slicedTickets.forEach(ticket => {
		// check if ticket details are rendered according to the ticket data passed
		const ticketElement = getByText(`${ticket.name} - ${formatDate(ticket.createdAt)}`)

		expect(ticketElement).toBeDefined()
	})
})

test('we should be able to delete tickets in cart menu clicking on trash icon', async () => {
	const delSpy = vi.spyOn(store, 'dispatch')
	const slicedTickets = tickets.slice(0, 2)
	rerender(<CartMenu tickets={slicedTickets} />)

	const trashIcons = getAllByTestId('delete-button')
	await user.click(trashIcons[0])
	expect(delSpy).toHaveBeenCalledWith(deleteTicket(slicedTickets[0].id))
})