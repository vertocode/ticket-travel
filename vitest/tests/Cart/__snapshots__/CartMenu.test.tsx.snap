// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot when there are no tickets 1`] = `
<div>
  <div
    class="cart-menu"
    data-testid="cart-menu"
  >
    <h3
      class="title"
    >
      Ingressos
    </h3>
    <div
      class="empty-cart-message"
      data-testid="empty-cart-message"
    >
      Não há ingressos no carrinho.
    </div>
  </div>
</div>
`;

exports[`should match snapshot when there are tickets 1`] = `
<div>
  <div
    class="cart-menu"
    data-testid="cart-menu"
  >
    <h3
      class="title"
    >
      Ingressos
    </h3>
    <div
      class="cart-ticket-details"
    >
      <img
        alt="Museu de Arte Moderna"
        data-nimg="1"
        decoding="async"
        height="70"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fi.postimg.cc%2FzD6RYYc7%2Fmuseu-de-arte.jpg&w=256&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.postimg.cc%2FzD6RYYc7%2Fmuseu-de-arte.jpg&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fi.postimg.cc%2FzD6RYYc7%2Fmuseu-de-arte.jpg&w=256&q=75 2x"
        style="color: transparent; object-fit: cover;"
        width="70"
      />
      <div
        class="cart-ticket-details-info"
      >
        <p
          class="title"
        >
          <span>
            Museu de Arte Moderna
             - 
            31/08/2022
          </span>
          <span
            class="delete-button"
            data-testid="delete-button"
          >
            <img
              alt="Deletar"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              src="/imgs/trash.svg"
              style="color: transparent;"
              width="20"
            />
          </span>
        </p>
        <p
          class="people"
        >
          1 Adulto: R$500,00 2 2 Crianças: R$234,33
        </p>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <p
          class="quantity-value"
        >
          <span>
            Qtd 
            1
          </span>
           
          <span>
            R$ 1.391,28
          </span>
        </p>
        <p
          class="sub-total"
        >
          <span>
            Subtotal
          </span>
           
          <span>
            R$ 1.391,28
          </span>
        </p>
      </div>
    </div>
    <div
      class="cart-ticket-details"
    >
      <img
        alt="Torre Eiffel"
        data-nimg="1"
        decoding="async"
        height="70"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fi.postimg.cc%2FJhFkqsSN%2Ftorre-eiffel.jpg&w=256&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.postimg.cc%2FJhFkqsSN%2Ftorre-eiffel.jpg&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fi.postimg.cc%2FJhFkqsSN%2Ftorre-eiffel.jpg&w=256&q=75 2x"
        style="color: transparent; object-fit: cover;"
        width="70"
      />
      <div
        class="cart-ticket-details-info"
      >
        <p
          class="title"
        >
          <span>
            Torre Eiffel
             - 
            31/08/2022
          </span>
          <span
            class="delete-button"
            data-testid="delete-button"
          >
            <img
              alt="Deletar"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              src="/imgs/trash.svg"
              style="color: transparent;"
              width="20"
            />
          </span>
        </p>
        <p
          class="people"
        >
          1 Adulto: R$500,00 2 2 Crianças: R$234,33
        </p>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <p
          class="quantity-value"
        >
          <span>
            Qtd 
            1
          </span>
           
          <span>
            R$ 1.391,28
          </span>
        </p>
        <p
          class="sub-total"
        >
          <span>
            Subtotal
          </span>
           
          <span>
            R$ 1.391,28
          </span>
        </p>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="cart-ticket-subtotal"
    >
      <p
        class="tickets-value"
      >
        <span>
          Ingressos
        </span>
        <span>
          R$ 4.702,56
        </span>
      </p>
      <p
        class="subtotal-value"
      >
        <span>
          Subtotal
        </span>
        <span>
          R$ 4.702,56
        </span>
      </p>
      <p
        class="discount-value"
      >
        <span>
          1X de 
          R$ 2.782,56
           com desconto de 
          <span
            class="percentage"
          >
            (
            41
            %)
          </span>
        </span>
        <span
          class="discount-value-currency"
        >
          -
          R$ 1.920,00
        </span>
      </p>
      <p
        class="portion-value"
      >
        <span>
          10X Sem juros de 
          R$ 278,26
        </span>
        <span>
          R$ 2.782,56
        </span>
      </p>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="cart-ticket-total"
    >
      <h3
        class="total"
      >
        <span>
          Valor total
        </span>
         
        <span
          class="value"
        >
          R$ 2.782,56
        </span>
      </h3>
    </div>
    <div
      class="check-out-button"
    >
      Ir para o check out
    </div>
  </div>
</div>
`;
